### PR TITLE
Added notebook demonstrating Featurestore

### PR DESCRIPTION
**Pull request by protorganizer**
Initial drat of notebook for featurestore, adjusted from internal version:

1. change the story
2. change the dataset
3. update env setup
4. use new sdk
5. rewrite the whole notebook with new example/data
6. style compliance

### CODE REVIEW

**Comment by ivanmkc** [comment]
Why is this in the unofficial folder? I thought the plan was to link to this from official docs?

File: ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb
Comment (ivanmkc) [inline ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb:1]: <p>Fix the link to point to the eventual URI.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (ivanmkc) [inline ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb:1]: <p>Don't think we need to show the uninstall step in our notebooks.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (ivanmkc) [inline ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb:1]: <p>We don't typically need to add this. I understand it's helpful but it's a little non-standard.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (ivanmkc) [inline ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb:1]: <p>Can this be done programmatically? This won't be able to pass CI testing in its current state.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

**Comment by ivanmkc** [comment]
There are a few cases where you use 'we' which needs fixing due to the guidelines from tech writers. I'll let your TW review take a look but please go over the notebook template once more to make sure you fix as much as possible.

File: ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb
Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>(global comment) "Feature Store" (the feature) is written as two words. "Featurestore" the resource is written as one.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>I believe we should number the steps (because we reference "step 1" down below: "# Output data set, as created in the step 1.")</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>I think it should be "...and specify the name of the table..."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>"data set" should be "dataset"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>I believe the genre feature is just a string (not a list). If so, I believe the description should be "The genre of the movie" (not genres)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>"python SDK" -&gt; "Python SDK"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='46'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>"each imported entity <em>must</em> have an ID on it, besides features; also, each entity can <em>optionally</em> have an timestamp on it" -&gt; "each imported entity <em>must</em> have an ID; also, each entity can <em>optionally</em> have a timestamp"</p><p>"located at this public-available <a href="https://pantheon.corp.google.com/storage/browser/cloud-samples-data/ai-platform-unified/datasets/featurestore;tab=objects?project=storage-samples&amp;prefix=&amp;forceOnObjectsSortingFiltering=false" target="_blank">bucket</a>." -&gt; "located at this publicly available <a href="https://pantheon.corp.google.com/storage/browser/cloud-samples-data/ai-platform-unified/datasets/featurestore;tab=objects?project=storage-samples&amp;prefix=&amp;forceOnObjectsSortingFiltering=false" target="_blank">bucket</a>."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='47'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>remove "currently support," which might cause uncertainty for users</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='48'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>Remove both commas from the "Note that..." sentence</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='60'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>To describe table 2, can we say something like "Table 2 joins the data from table 1 and data from the featurestore." Is that accurate?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='65'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>Can liked genres change over time? If so, maybe change to "...feature store returns <code>age=54</code> along with alice's liked genres at the time in the training data because that is the..."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='66'/>

File: ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb
Comment (protorganizer) [reply ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb]: <p>If a user has installed old version, this NB would not work.</p>

Comment (protorganizer) [reply ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb]: <p>removed</p>

Comment (protorganizer) [reply ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb]: <p>Do we have a BigQuery NB example? I can try to look for a programmatic way from there.</p>

File: ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb
Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>Are you suggesting we use Featurestore across the NB?</p>

Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>added</p>

Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>fixed</p>

Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>In this NB, I use singular to simplify the data generation. But in the original dataset [1], it is plural because one movie can be categorized as more than one genre. Do you think we can let this slide? </p><p> </p><p>[1] https://files.grouplens.org/datasets/movielens/ml-25m-README.html</p>

Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>fixed</p>

Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>That is accurate. Added</p>

Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>Correct, all other feature values can all be time-variant. </p><p> </p><p>Appended this to explicitly speak it out.</p>

Comment (ey00) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: It's context dependent. If we're referring to the feature, it's Feature Store. If we're referring to the resource (the top-level container), it's featurestore.

Comment (ey00) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: I see. If I understand correctly, we're treating the column value as a single string even in instances where there are multiple genres, ex: "Comedy|Romance"? Is this a best practice? If not, what should users do instead? Maybe we should add something like "Note: For simplicity, this notebook treats the genres column values as a string even though a value might include multiple genres. For example, a movie might have a `genres` feature value of "Comedy|Romance," which will be imported as a single string. In real-world scenarios, you might set the `genres` feature type to an array and transform the column to an array before importing the data."

Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>fixed according to the context, ptal</p>

Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>In this specific notebook, one movie is always categorized as one genre: I fake the data so that each movie only has one genre and define the schema as such. So no "Comedy|Romance": I just pick one of them in such cases. Again just for simplicity purpose and the data is purely faked, according to the movie recommendation story.</p><p> </p><p>In the future, we can re-generate dataset and re-write the notebook accordingly if we decide to make the data more accurate/authentic.</p>

File: ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb
Comment (protorganizer) [reply ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb]: <p>changed to using bq clients</p>

File: ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb
Comment (amygdala) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:882]: looks like this bucket should be 'cloud-samples-data-us-central1'

File: ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb
Comment (amygdala) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1031]: I feel like it might also be useful to show an example of how to unpack the result returned.  E.g. it would be something like 
`result.entity_view.data[2].value.string_array_value.values` to get the list of liked genres... or is there a better way?

File: ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb
Comment (amygdala) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:993]: a high-level comment: if we add an example of actually training a model, as discussed, then I think the flow would work better to first show the batch serving case, then train the model using the batch data, then the online serving API case, showing how to construct a prediction request. 
[For simplicity I don't think the model would have to be saved and deployed; it could just be an in-notebook object]

File: ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb
Comment (amygdala) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:446]: Unless I did something goofy I don't think INPUT_CSV_FILE is the right file?
```
gsutil cat "gs://cloud-aiplatform-featurestore-us-central1/read_entity_instance.csv"|head -20 
bikes,stations,timestamp
097G,3841,2019-10-27T08:28:14Z
057G,2565,2019-10-19T08:58:56Z
2266,2707,2019-10-15T09:24:11Z
```

**Comment by amygdala** [comment]
high-level comment: As discussed in that email thread, could we broaden this example show use of embeddings as well?

File: ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb
Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb]: IIRC, The reason why we put online serving ahead of batch, in this standalone notebook, is that online serving is the most-asked feature from customer.

But I agree that a data-ingestion + model training + model serving version tells a wonderful e2e story. The only question is if we consider the tight timeline before I/O, I think most likely we will use this as a standalone notebook.

File: ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb
Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb]: this is a bug, fixed

File: ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb
Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb]: > typo: groud--> ground
fixed
>in that more generally you might be including not only the labels but other info too, such as (for this use case) things like referrer URL, how long spent watching, etc...
Are you suggesting that the prediction output should be more than just a yes/no label? Ideally, yes, but it's not supported yet. Actually even if the boolean label is not explicitly supported yet. Please refer to the tables down below fore more explaination.
> It would be helpful to add just a sentence on that 
It was mentioned in the env setup section, when the DESTINATION_TABLE_URI is defined. I agree that it might be too far away from where it's used but it is recommended by the template that we put all const def together in that section.
Anyways, added one more sentence as a reminder.

File: ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb
Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>"Google cloud" -&gt; "Google Cloud" (for branding)</p><p> </p><p>fixed typo and clarified that it's a dataset</p><p>"...the movie recommentation as an example..." -&gt; "...a move recommendation dataset as an example..."</p><p> </p><p>"into the featurestore service." -&gt; "into Feature Store." OR "into a featurestore." (if you want to refer to the featurestore resource)</p><p> </p><p>"How to serve online prediction requests..." (make "request" plural)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>There's a "TODO" on step 3, should that be removed?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>"Feature Store organizes data with the following..." (capitalize Store)</p><p> </p><p>"you will create a featurestore" (use featurestore one word)</p><p> </p><p>"The Users entity type has the age, gender, and like genres features." (add "type" "the" and "and")</p><p> </p><p>"The Movies entity type has the genres and average rating features." (similar change as above)</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>"...into the featurestore."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='50'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>"It's designed for a latency-sensitive service..."</p><p> </p><p>Something is missing from the the last sentence: "For example, if you have a model to predict if a user is going to watch" Do you mean, "For example, for a movie service, you might want to quickly shows movies that the current user would most likely watch by using online predictions." or something like that?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='55'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>"By default, the API will return the &nbsp;latest value of each feature, meaning the feature values with the most recent &nbsp;timestamp."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='56'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>"The task is to prepare a training dataset..." (remove "for")</p><p> </p><p>"...from Feature Store according to the..." (I think we're referring to Feature Store the service here)</p><p> </p><p>"...feature values from the featurestore..." (add "the")</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='63'/>

Comment (ey00) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>"...in a BigQuery table." (branding fix)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='65'/>

File: ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb
Comment (diemtvu) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:530]: What is the reason for having timestamp as postfix? I wonder if it led people think FS is created dynamically.

Comment (protorganizer) [reply ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:530]: This is to temporarily give it a unique name, because if for some reason the tests failed, the created featurestore will not be cleaned up. Next time we run the test, we will get a "featurestore already exist" error.

This can be removed after we pass the test and merge it with upstream, when we will no longer need to run it in the test env.

**Review by ey00 — APPROVED** [review]
just one minor fix

**Review by ey00 — APPROVED** [review]

File: ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb
Comment (ivanmkc) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>Please use the injected REGION. See https://docs.google.com/document/d/1MjbMDrm4GvfLJfqnPmdTUS8dueKdRAUsn5bmm2aCoZo/edit?resourcekey=0-LE_XfwNqTclOKG8hbFnxrQ#bookmark=id.61pzrvibw4mf</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (ivanmkc) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>Please use injected REGION: https://docs.google.com/document/d/1MjbMDrm4GvfLJfqnPmdTUS8dueKdRAUsn5bmm2aCoZo/edit?resourcekey=0-LE_XfwNqTclOKG8hbFnxrQ#bookmark=id.61pzrvibw4mf</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (ivanmkc) [inline ai-platform-unified/notebooks/official/featurestore_movie_prediction.ipynb:1]: <p>I think you have to delete the BQ dataset as well.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/ai-platform-samples/pull/283/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='71'/>

File: ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb
Comment (ivanmkc) [reply ai-platform-unified/notebooks/unofficial/featurestore/featurestore_movie_prediction.ipynb]: Oh does `pip install -U` not override the version?

**Review by ivanmkc — APPROVED** [review]
LGTM

**Pull request merged**
